src/webgpu/shader/validation/uniformity/uniformity.spec.ts

```diff
@@ -354,3 +355,900 @@ fn main(@builtin(${t.params.builtin}) p : ${t.params.type}) {
+    typename: `block`,
+    typedecl: `struct block {
+      x : u32,
+      y : u32
+    }`,
+    assignment: `x.x = uniform_value[0].x;`,
+    cond: `x.x > 0`,
+    uniform: `init`,
+  },
+  partial_assignment_nonuniform: {
+    typename: `block`,
+    typedecl: `struct block {
+      x : u32,
+      y : u32
+    }`,
+    assignment: `x.x = nonuniform_value[0].x;`,
+    cond: `x.x > 0`,
+    uniform: `never`,
+  },
+  partial_assignment_all_members_uniform: {
+    typename: `block`,
+    typedecl: `struct block {
+      x : u32,
+      y : u32
+    }`,
+    assignment: `x.x = uniform_value[0].x;
+    x.y = uniform_value[1].y;`,
+    cond: `x.x > 0`,
+    uniform: `init`,
+  },
+  partial_assignment_all_members_nonuniform: {
+    typename: `block`,
+    typedecl: `struct block {
+      x : u32,
+      y : u32
+    }`,
+    assignment: `x.x = nonuniform_value[0].x;
+    x.y = uniform_value[0].x;`,
+    cond: `x.x > 0`,
+    uniform: `never`,
+  },
+  partial_assignment_single_element_struct_uniform: {
+    typename: `block`,
+    typedecl: `struct block {
+      x : u32
+    }`,
+    assignment: `x.x = uniform_value[0].x;`,
+    cond: `x.x > 0`,
+    uniform: `init`,
+  },
+  partial_assignment_single_element_struct_nonuniform: {
+    typename: `block`,
+    typedecl: `struct block {
+      x : u32
+    }`,
+    assignment: `x.x = nonuniform_value[0].x;`,
+    cond: `x.x > 0`,
+    uniform: `never`,
+  },
+  partial_assignment_single_element_array_uniform: {
+    typename: `array<u32, 1>`,
+    typedecl: ``,
+    assignment: `x[0] = uniform_value[0][0];`,
+    cond: `x[0] > 0`,
+    uniform: `init`,
+  },
+  partial_assignment_single_element_array_nonuniform: {
+    typename: `array<u32, 1>`,
+    typedecl: ``,
+    assignment: `x[0] = nonuniform_value[0][0];`,
+    cond: `x[0] > 0`,
+    uniform: `never`,
+  },
+  nested1: {
+    typename: `block`,
+    typedecl: `struct block {
+      x : u32,
+      y : u32
+    }`,
+    assignment: `for (; uniform_cond; ) {
+      if uniform_cond {
+        x = uniform_value[0];
+        break;
+        x.y = nonuniform_value[0].y;
+      } else {
+        if uniform_cond {
+          continue;
+        }
+        x = uniform_value[1];
+      }
+    }`,
+    cond: `x.x > 0`,
+    uniform: `init`,
+  },
+  nested2: {
+    typename: `block`,
+    typedecl: `struct block {
+      x : u32,
+      y : u32
+    }`,
+    assignment: `for (; uniform_cond; ) {
+      if uniform_cond {
+        x = uniform_value[0];
+        break;
+        x.y = nonuniform_value[0].y;
+      } else {
+        if nonuniform_cond {
+          continue;
+        }
+        x = uniform_value[1];
+      }
+    }`,
+    cond: `x.x > 0`,
+    uniform: `never`,
+  },
+};
+
+g.test('function_variables')
+  .desc(`Test uniformity of function variables`)
+  .params(u =>
+    u
+      .combine('case', keysOf(kFuncVarCases))
+      .combine('init', ['none', 'uniform', 'nonuniform'] as const)
+  )
+  .fn(t => {
+    const code = `
```

Comment:
nit: Since you are pulling multiple fields from a single entry in kFuncCases below, possibly pull the element out one and save it to a local variable `const func_case = kFuncVarCases[t.params.case];`